server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -74,7 +79,23 @@ public synchronized void onInit(List<? extends IndexCommit> commits) throws IOEx
     }
 
     @Override
-    public synchronized void onCommit(List<? extends IndexCommit> commits) throws IOException {
+    public void onCommit(List<? extends IndexCommit> commits) throws IOException {
+        final IndexCommit safeCommit = updateCommitsAndRetentionPolicy(commits);
+        final long localCheckpointOfSafeCommit = Long.parseLong(safeCommit.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY));
+        final long docCountOfSafeCommit = getDocCountOfSafeCommit();
+        minimumReasonableRetainedSeqNo = localCheckpointOfSafeCommit + 1
```

Comment:
what was the reason to compute this value here and not in ReplicationTracker? This class could e.g. return a pre-computed `safeCommitInfo` that contains local checkpoint + number of docs.
The actual computation could then be done in replication tracker, where I think that logic fits more naturally (and avoids the slightly odd `minimumReasonableRetainedSeqNo` naming that we have here).

Reply:
It didn't really seem worth the extra plumbing to move this single expression, particularly since `CombinedDeletionPolicy` also has some bearing on history retention. It didn't seem any more natural to have it in `ReplicationTracker`. But I've done this in 13a4c38.